gt/main.c

```diff
@@ -219,6 +262,40 @@ lookup_policy_decision(struct gt_packet_headers *pkt_info,
 	return 0;
 }
 
+static int
+lookup_frag_punish_policy_decision(struct gt_packet_headers *pkt_info,
+	struct ggu_policy *policy, struct gt_instance *instance)
+{
+	policy->flow.proto = pkt_info->inner_ip_ver;
+	if (pkt_info->inner_ip_ver == ETHER_TYPE_IPv4) {
+		struct ipv4_hdr *ip4_hdr = pkt_info->inner_l3_hdr;
+
+		policy->flow.f.v4.src = ip4_hdr->src_addr;
+		policy->flow.f.v4.dst = ip4_hdr->dst_addr;
+	} else if (likely(pkt_info->inner_ip_ver == ETHER_TYPE_IPv6)) {
+		struct ipv6_hdr *ip6_hdr = pkt_info->inner_l3_hdr;
+
+		rte_memcpy(policy->flow.f.v6.src, ip6_hdr->src_addr,
+			sizeof(policy->flow.f.v6.src));
+		rte_memcpy(policy->flow.f.v6.dst, ip6_hdr->dst_addr,
+			sizeof(policy->flow.f.v6.dst));
+	} else
+		rte_panic("Unexpected condition: gt block at lcore %u lookups policy decision for an non-IP packet!\n",
+			rte_lcore_id());
```

Comment:
We cannot use `rte_panic()` here, instead, we need to fail gracefully. Otherwise, if an internal attacker, or an attacker that can inject packets in the GK-GT channel, can easily bring Grantor servers down. Let's just log an error message and return an error as we do with `lua_pcall()` below.